Add SaaS Application support

ouranos opened the pull request:
Closes #1226

Added in cloudflare-go: https://github.com/cloudflare/cloudflare-go/pull/900

I've tested in our environment which had a custom fork of the SDK previously and the plan is not showing any difference when using the new SDK. Opening a draft PR to get some feedback and I'll do more testing in the meantime.

I need a bit of help handling the `domain` attribute. From my research (before the doc was available), it seems that it is not required for SaaS applications and will be computed for them (eg: `https://<organisation>.cloudflareaccess.com/cdn-cgi/access/sso/saml/<uuid>`.
The updated documentation seems to confirm it: https://api.cloudflare.com/#access-applications-properties

jacobbednarz commented:
@Justin-Holmes can i get your eyes on this too please?

jacobbednarz commented on internal/provider/resource_cloudflare_access_application_test.go:
```diff
@@ -103,6 +105,34 @@ func TestAccCloudflareAccessApplication_WithCORS(t *testing.T) {
+		CheckDestroy:      testAccCheckCloudflareAccessApplicationDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCloudflareAccessApplicationConfigWithSaas(rnd, accountID),
```
accountID is undefined here. you'll need `accountID := os.Getenv("CLOUDFLARE_ACCOUNT_ID")` to use it.

ouranos replied:
I copied the syntax from other tests in this file (eg: `TestAccCloudflareAccessApplication_BasicAccount` & `TestAccCloudflareAccessApplication_WithADefinedIdps`) and they don't seem to define it either. Is that an issue too?

jacobbednarz replied:
looks like `resource_cloudflare_access_group_test.go` is defining `accountID` globally and the others are relying on it. definitely don't rely on this and instead define it at the test level.

ouranos replied:
Understood, done.

jacobbednarz commented on internal/provider/schema_cloudflare_access_application.go:
```diff
@@ -121,6 +123,28 @@ func resourceCloudflareAccessApplicationSchema() map[string]*schema.Schema {
 				},
 			},
 		},
+		"saas_app": {
```
all these fields need to be documented with `Description` fields to generate the documentation

ouranos replied:
Done

ouranos commented on internal/provider/resource_cloudflare_access_application.go:
```diff
@@ -64,6 +64,11 @@ func resourceCloudflareAccessApplicationCreate(ctx context.Context, d *schema.Re
 
+	if _, ok := d.GetOk("saas_app"); ok {
+		saasApp := convertSaasSchemaToStruct(d)
+		newAccessApplication.SaasApplication = saasApp
```
Done

Justin-Holmes commented:
Looks great! And you're correct @ouranos, the `domain` is not required for `saas` applications and is generated for you on creation. 

Also, the `custom_attributes` field is not used often so I'm cool omitting that for now.

ouranos commented:
Thanks, 

> the `domain` is not required for `saas` applications and is generated for you on creation.

So this is the correct way then?
```diff
"domain": {
    Type:     schema.TypeString,
-     Required: true,
+     Optional: true,
+     Computed: true, 
    ...
}
```

Let me know if you are happy with it or if it needs more change and I'll rebase/clean up/squash.

jacobbednarz commented:
> So this is the correct way then?

`Computed: true` is fine. `Optional` implies the end user is also able to set it.

jacobbednarz commented on internal/provider/schema_cloudflare_access_application.go:
```diff
@@ -37,15 +37,17 @@ func resourceCloudflareAccessApplicationSchema() map[string]*schema.Schema {
-			Type:        schema.TypeString,
-			Required:    true,
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true, // REVIEW
+			// Required:    true,
```
```suggestion
			Computed: true,
```

ouranos replied:
We still need to be able to set it for `self_hosted`, wouldn't that break it?

jacobbednarz replied:
that would; if we need to set it, this is fine.

jacobbednarz commented on internal/provider/schema_cloudflare_access_application.go:
```diff
@@ -121,6 +123,32 @@ func resourceCloudflareAccessApplicationSchema() map[string]*schema.Schema {
 			},
 		},
+		"saas_app": {
+			Type:        schema.TypeList,
+			Optional:    true,
+			Description: "SaaS configuration for the Access Application. See below for reference structure.",
```
```suggestion
			Type:        schema.TypeList,
			Optional:    true,
			MaxItems:    1,
			Description: "SaaS configuration for the Access Application.",
```

jacobbednarz commented on internal/provider/resource_cloudflare_access_application_test.go:
```diff
@@ -103,6 +105,34 @@ func TestAccCloudflareAccessApplication_WithCORS(t *testing.T) {
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccessAccPreCheck(t)
```
you need both of these here (sets up the mock client)

```suggestion
            testAccPreCheck(t)
			testAccessAccPreCheck(t)
```

ouranos commented:
> > So this is the correct way then?
> 
> `Computed: true` is fine. `Optional` implies the end user is also able to set it.

It's computed for SaaS apps (and returns a Cloudflare generated domain) but it's required for the other types (eg: self-hosted)

ouranos commented:
> don't forget to run `make docs` to get the documentation updated here. you'll need to fix the tests as well now that the field is computed.

- Squashed all previous commits together
- Addressed the new review comments
- Generated the docs

So I think the only remaining points are:
- `domain`: it's `required` for `self_hosted` but computed for `saas`
- acceptance tests: I only ran `make test` so far and I'm now trying to set up an environment to run `make testacc`.  We're only using Zero-Trust with Saas apps so we don't have any domains/zones set up. Looks like I can only add root domains on the free plan (we do have a paid plan for ZeroTrust). I'm going to try to set up one of our unused test domains.

jacobbednarz commented on internal/provider/schema_cloudflare_access_application.go:
```diff
@@ -37,15 +37,17 @@ func resourceCloudflareAccessApplicationSchema() map[string]*schema.Schema {
 			Default:      "self_hosted",
-			ValidateFunc: validation.StringInSlice([]string{"self_hosted", "ssh", "vnc", "file"}, false),
+			ValidateFunc: validation.StringInSlice([]string{"self_hosted", "saas", "ssh", "vnc", "file"}, false),
 			Description:  fmt.Sprintf("The application type. %s", renderAvailableDocumentationValuesStringSlice([]string{"self_hosted", "ssh", "vnc", "file"})),
```
need to add `saas` here too
```suggestion
			Description:  fmt.Sprintf("The application type. %s", renderAvailableDocumentationValuesStringSlice([]string{"self_hosted", "saas", "ssh", "vnc", "file"})),
```

jacobbednarz commented on internal/provider/schema_cloudflare_access_application.go:
```diff
@@ -121,6 +123,33 @@ func resourceCloudflareAccessApplicationSchema() map[string]*schema.Schema {
+			Type:        schema.TypeList,
+			Optional:    true,
+			MaxItems:    1,
+			Description: "SaaS configuration for the Access Application. See below for reference structure.",
```
```suggestion
			Description: "SaaS configuration for the Access Application.",
```

jacobbednarz commented on internal/provider/resource_cloudflare_access_application_test.go:
```diff
@@ -103,6 +105,36 @@ func TestAccCloudflareAccessApplication_WithCORS(t *testing.T) {
 	})
 }
 
+func TestAccCloudflareAccessApplication_WithSass(t *testing.T) {
```
```suggestion
func TestAccCloudflareAccessApplication_WithSaas(t *testing.T) {
```

jacobbednarz commented on internal/provider/resource_cloudflare_access_application_test.go:
```diff
@@ -103,6 +105,36 @@ func TestAccCloudflareAccessApplication_WithCORS(t *testing.T) {
+					resource.TestCheckResourceAttr(name, "type", "saas"),
+					resource.TestCheckResourceAttr(name, "session_duration", "24h"),
+					resource.TestCheckResourceAttr(name, "saas_app.#", "1"),
+					resource.TestCheckResourceAttr(name, "saas_app.0.sp_entity_id", "1"),
+					resource.TestCheckResourceAttr(name, "saas_app.0.consumer_service_url", "1"),
+					resource.TestCheckResourceAttr(name, "saas_app.0.name_id_format", "1"),
```
need to fix these assertions

```suggestion
					resource.TestCheckResourceAttr(name, "saas_app.0.sp_entity_id", "saas-app.example"),
					resource.TestCheckResourceAttr(name, "saas_app.0.consumer_service_url", "https://saas-app.example/sso/saml/consume"),
					resource.TestCheckResourceAttr(name, "saas_app.0.name_id_format", "email"),
```

jacobbednarz replied:
if you fix these, the acceptance tests all pass

ouranos commented:
> if you fix these, the acceptance tests all pass

Great thanks 🙏 

I've:
- applied all suggested changes
- cleaned up the comments regarding `domain`
- rebased from latest `master` & squashed

Let me know if I need to do anything else

ouranos commented:
> did you re-run `make docs`? that should be the only thing.

Nope 🤦‍♂️  
Done now. It fixes doc for #1792 too

jacobbednarz commented:
cool! I'll run the acceptance tests again in the morning and we should be good, thanks!

p.s for future PRs if you check the box to allow maintainers to make changes, we can limit the back and forth as I can provide a patch within the branch for you.
